### PR TITLE
Rewrite the "Authentication" doc for clarity and flow.

### PR DESCRIPTION
**Pull request by sergetron**
# Changes

Rewrites the "Authentication" doc for clarity and flow.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [X] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [X] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```release-note
NONE
```

### CODE REVIEW

File: docs/auth.md [hunk 1]
```diff
@@ -4,35 +4,225 @@ linkTitle: "Authentication"
 weight: 7
 ---
 -->
-# Authentication
+# Authentication at Run Time
+
+This document describes how Tekton handles authentication when executing
+`TaskRuns` and `PipelineRuns`. Since authentication concepts and processes
+apply to both of those entities in the same manner, this document collectively
+refers to `TaskRuns` and `PipelineRuns` as `Runs` for the sake of brevity.
```
Comment (dlorenc) [inline docs/auth.md:12]: This might get confusing with the work @ImJasonH is doing, which introduces a new higher-level concept called "Runs". :(

I think it's fine for now, but maybe worth thinking about.

File: docs/auth.md [hunk 2]
```diff
@@ -4,35 +4,225 @@ linkTitle: "Authentication"
...
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  annotations:
+    tekton.dev/git-0: github.com
+type: kubernetes.io/ssh-auth
+data:
+  ssh-privatekey: <base64-encoded>
+  # This is non-standard, but its use is encouraged to make this more secure.
```
Comment (dlorenc) [inline docs/auth.md:104]: The known_hosts section is probably better to remove from this example. It isn't related to the rest of the way the Secret is setup, and might be distracting.

File: docs/auth.md [hunk 3]
```diff
@@ -43,26 +233,23 @@ aggregates them into their respective files in `$HOME`.
        tekton.dev/git-0: github.com # Described below
    type: kubernetes.io/ssh-auth
    data:
```
Comment (dlorenc) [inline docs/auth.md:235]: We use stringData above to avoid requiring users to base64 encode the data, it would probably be better to use it here as well.

File: docs/auth.md [hunk 4]
```diff
@@ -43,26 +233,23 @@ aggregates them into their respective files in `$HOME`.
...
-   [Guiding Credential Selection](#guiding-credential-selection) below for more
-   information.
+   In the above example, the value for `tekton.dev/git-0` specifies the URL for which Tekton will use this `Secret`,
+   as described in [Understanding credential selection](#understanding-credential-selection).
+
+1. Generate the `ssh-privatekey` value. For example:
 
-1. Generate the value of `ssh-privatekey` by copying the value of (for example)
-   `cat ~/.ssh/id_rsa | base64`.
+   `cat ~/.ssh/id_rsa | base64`
```
Comment (dlorenc) [inline docs/auth.md:248]: See above about stringData and dropping the need to base64 encode

File: docs/auth.md [hunk 5]
```diff
@@ -258,61 +371,63 @@ credentials are then used to authenticate when retrieving any
...
 
    ```shell
    kubectl apply --filename secret.yaml serviceaccount.yaml run.yaml
    ```
 
-When the `Run` executes, before steps execute, a `~/.docker/config.json` will be
-generated containing the credentials configured in the `Secret`, and these
-credentials are then used to authenticate when retrieving any
-`PipelineResources`.
+## Configuring `docker*` authentication for Docker
```
Comment (dlorenc) [inline docs/auth.md:409]: Why the * in docker*?

Comment (sergetron) [reply docs/auth.md:409]: Because the section covers two Secrets called `dockercfg` and `dockerconfigjson`, hence `docker*`. Writing both names out would make the section title really long.

File: docs/auth.md (hunk 1 above)
Comment (sergetron) [reply docs/auth.md:12]: When will Jason's "Runs" be ready? We can update the doc then.

File: docs/auth.md (hunk 2 above)
Comment (sergetron) [reply docs/auth.md:104]: Deleted the `known_hosts` field.

File: docs/auth.md (hunk 3 above)
Comment (sergetron) [reply docs/auth.md:235]: Which lines should I change here and to what?

File: docs/auth.md (hunk 4 above)
Comment (sergetron) [reply docs/auth.md:248]: Which lines should I change here and to what?

Comment (dlorenc) [reply docs/auth.md:247]: I'll try to comment on the lines.

File: docs/auth.md [hunk 6]
```diff
@@ -43,26 +232,23 @@ aggregates them into their respective files in `$HOME`.
        tekton.dev/git-0: github.com # Described below
    type: kubernetes.io/ssh-auth
    data:
```
Comment (dlorenc) [inline docs/auth.md:234]: Change "data" to "stringData"

File: docs/auth.md [hunk 7]
```diff
@@ -43,26 +232,23 @@ aggregates them into their respective files in `$HOME`.
...
-     ssh-privatekey: <base64 encoded>
+     ssh-privatekey: <base64-encoded>
```
Comment (dlorenc) [inline docs/auth.md:235]: Change base64-encoded to just private-key

File: docs/auth.md [hunk 8]
```diff
@@ -43,26 +232,23 @@ aggregates them into their respective files in `$HOME`.
...
      # This is non-standard, but its use is encouraged to make this more secure.
      # If it is not provided then the git server's public key will be requested
      # when the repo is first fetched.
-     known_hosts: <base64 encoded>
+     known_hosts: <base64-encoded>
```
Comment (dlorenc) [inline docs/auth.md:239]: Change base64-encoded to just known_hosts

File: docs/auth.md (hunk 7 above)
Comment (sergetron) [reply docs/auth.md:235]: Done.

File: docs/auth.md (hunk 8 above)
Comment (sergetron) [reply docs/auth.md:239]: Done.

File: docs/auth.md (hunk 6 above)
Comment (sergetron) [reply docs/auth.md:234]: Done.

File: docs/auth.md (hunk 3 above)
Comment (sergetron) [reply docs/auth.md:234]: All fixed.

**Comment by sergetron** [comment]
All comments addressed! Thanks for the review. :)

File: docs/auth.md [hunk 9]
```diff
@@ -42,27 +231,24 @@ aggregates them into their respective files in `$HOME`.
...
-   [Guiding Credential Selection](#guiding-credential-selection) below for more
-   information.
+   In the above example, the value for `tekton.dev/git-0` specifies the URL for which Tekton will use this `Secret`,
+   as described in [Understanding credential selection](#understanding-credential-selection).
+
+1. Generate the `ssh-privatekey` value. For example:
 
-1. Generate the value of `ssh-privatekey` by copying the value of (for example)
-   `cat ~/.ssh/id_rsa | base64`.
+   `cat ~/.ssh/id_rsa | base64`
```
Comment (dlorenc) [inline docs/auth.md:247]: D'oh, missed one. We also need to drop the "| base64" here.

File: docs/auth.md (hunk 1 above)
Comment (sergetron) [reply docs/auth.md:12]: Leaving it as-is, we'll revisit once the Jason "Runs" are out. :)

File: docs/auth.md (hunk 9 above)
Comment (sergetron) [reply docs/auth.md:247]: No worries. :) Fixed!

**Comment by sergetron** [comment]
Can we get LGTM + APPROVAL please?

File: docs/auth.md [hunk 10]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+convert them as follows:
+
+- For Git, convert into a `.gitconfig` file.
+- For Docker, convert into a `~/.docker/config.json` file.
+
+Additionally, even though each of these entities supports multiple credentials for multiple
+domains, Tekton must first blend those credentials into a single canonical keyring.
+
+To consume these `Secrets`, Tekton performs credential initialization within every `Pod` it instantiates,
+before retrieving any associated `PipelineResources`. During credential initialization, Tekton accesses
```
Comment (vdemeester) [inline docs/auth.md:63]: nit: this is not related to `PipelineResources`. Even without any `PipelineResource` the *credential initialijation* will happen.

File: docs/auth.md [hunk 11]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+each `Secret` associated with the `Run` and aggregates all of those `Secrets` into their respective
+files in `$HOME/tekton/home`.
```
Comment (vdemeester) [inline docs/auth.md:65]: We are writing the credentials in `/tekton/creds` and those are *linked* or *copied* depending on some configuration.
cc @sbwsg

File: docs/auth.md [hunk 12]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+## Limiting `Secret` access to specific `Steps`
 
-Secrets of these types can be made available to the `Run` by attaching them to
-the `ServiceAccount` as which it runs.
+As described earlier in this document, Tekton stores supported `Secrets` in
+`$HOME/tekton/home` and makes them available to all `Steps` within a `Task`. 
 
-## Exposing credentials
+If you want to limit a `Secret` to only be accessible to specific `Steps` but not
+others, you must explicitly specify a `Volume` within the `Secret` definition and
```
Comment (sbwsg) [inline docs/auth.md:133]: Suggest "specify a `Volume` using the `Secret` definition".  The volume isn't defined in the Secret.

File: docs/auth.md (hunk 10 above)
Comment (sergetron) [reply docs/auth.md:63]: Changed to "[...] before executing any `Steps` in the `Run`."

File: docs/auth.md (hunk 11 above)
Comment (sergetron) [reply docs/auth.md:65]: Reworded.

File: docs/auth.md [hunk 13]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+				<code>kubernetes.io/ssh-auth</code>
+			</td>
+			<td><code>kubernetes.io/basic-auth</code><br>
+				<code>kubernetes.io/dockercfg</code><br>
+				<code>kubernetes.io/dockerconfigjson</code>
+			</td>
+	</tbody>
+</table>	
+
+Since neither Git nor Docker can consume these `Secrets` in their native format, Tekton must first
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:53]: Reworded.

File: docs/auth.md [hunk 14]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+convert them as follows:
+
+- For Git, convert into a `.gitconfig` file.
+- For Docker, convert into a `~/.docker/config.json` file.
+
+Additionally, even though each of these entities supports multiple credentials for multiple
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:59]: Reworded.

File: docs/auth.md [hunk 15]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+see [`authenticating-git-commands`](../examples/v1beta1/taskruns/authenticating-git-commands.yaml).
 
-- `kubernetes.io/basic-auth`
-- `kubernetes.io/ssh-auth`
+## Limiting `Secret` access to specific `Steps`
 
-Secrets of these types can be made available to the `Run` by attaching them to
-the `ServiceAccount` as which it runs.
+As described earlier in this document, Tekton stores supported `Secrets` in
+`$HOME/tekton/home` and makes them available to all `Steps` within a `Task`.
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:130]: Fixed.

File: docs/auth.md [hunk 16]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+		<tr>
+			<td><code>kubernetes.io/basic-auth</code><br>
+				<code>kubernetes.io/ssh-auth</code>
+			</td>
+			<td><code>kubernetes.io/basic-auth</code><br>
+				<code>kubernetes.io/dockercfg</code><br>
+				<code>kubernetes.io/dockerconfigjson</code>
+			</td>
+	</tbody>
+</table>
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:51]: Added!

File: docs/auth.md [hunk 17]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+  annotations:
+    tekton.dev/git-0: github.com
+type: kubernetes.io/ssh-auth
+stringData:
+  ssh-privatekey: <private-key>
+  # This is non-standard, but its use is encouraged to make this more secure.
+  # Omitting this results in the server's public key being blindly accepted.
+```
+
+## Authenticating as a non-root user
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:108]: Reworded.

File: docs/auth.md (hunk 12 above)
Comment (sergetron) [reply docs/auth.md:133]: Reworded.

File: docs/auth.md [hunk 18]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+
+In certain scenarios you might need to authenticate as a non-root user. For example:
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:110]: Reworded.

File: docs/auth.md [hunk 19]
```diff
@@ -4,35 +4,224 @@ linkTitle: "Authentication"
...
+
+To consume these `Secrets`, Tekton performs credential initialization within every `Pod` it instantiates,
+before retrieving any associated `PipelineResources`. During credential initialization, Tekton accesses
+each `Secret` associated with the `Run` and aggregates all of those `Secrets` into their respective
+files in `$HOME/tekton/home`.
+
+## Understanding credential selection
+
+A `Run` might require multiple types of authentication. For example, a `Run` might require access to
+multiple private Git and Docker repositories. You can use credential annotation keys to direct Tekton to
```
[the remark this answers is not in the document]
Comment (sergetron) [reply docs/auth.md:70]: Reworded.

**Comment by sergetron** [comment]
All comments addressed. Thank you Scott and Vincent for the thorough review. :)

File: docs/auth.md [hunk 20]
```diff
@@ -4,35 +4,226 @@ linkTitle: "Authentication"
...
+- The `Steps` in your `Task` define a non-root `securityContext`.
+- Your `Task` specifies a global non-root `securityContext` that applies to all `Steps` in the `Task`.
+
+The following are considerations for executing `Runs` as a non-root user:
+
+- `ssh-auth` for Git requires the user to have a valid home directory configured in `/etc/passwd`.
+  Specifying a UID that has no valid home directory results in authentication failure.
+- Since SSH authentication ignores the `$HOME` environment variable, you must either move or symlink
+  the appropriate `Secret` files from the `$HOME` directory defined by Tekton (`/tekton/home`) to
+  the the non-root user's valid home directory to use SSH authentication for either Git or Docker.
```
Comment (sbwsg) [inline docs/auth.md:124]: v minor nit: `the the` -> `the`

**Comment by sergetron** [comment]
Can we have an LGTM please?
